Update to Angular 19 with rework

joematthews opened the pull request:
update to angular 19 with rework

view new repo/README.md: https://github.com/joematthews/extreme-angular/tree/update-to-angular-19-with-rework

Draft changelog:

# What's Changed

---

## Configuration Changes

- Update to Angular 19 (and rework config to match `@angular/cli` generated output)
- Switch all `.js` config files to [ECMAScript Modules](https://nodejs.org/api/esm.html#modules-ecmascript-modules) (special thanks to @replete for this change)
  - Add `"type": "module",` to `package.json`
- Add new scripts to `package.json`:
  - `lint:spelling`: check spelling in all files
  - `lint:tsc:app`: check for Typescript errors in application files
  - `lint:tsc:spec`: check for Typescript errors in spec/test files
  - `lint:format`: silently check formatting in all files
  - `lint:all`: run all lint scripts
  - `ci:all`: run `lint:all` and `test:ci`
  - `shove`: quickly add, commit, and push all files without checks or tests (for quick back up in case of emergency)
- Add `on-pull-request.yml` and `validate-code.yml` Github Action workflows.
  - Run `ci:all` script in `pre-push` hook and `validate-code.yml` workflow.
- Add `/.husky/_` to .gitignore to prevent false negatives when running `lint:format`
- Add dictionaries to `.cspell.json` to help reduce false negatives
- Print 'Example conventional-commit messages' on commitlint failure
- Adjust `.editorconfig` to make it easier to match indent_size with Prettier tabWidth
- Remove `@angular/material`, `@angular/ssr`, and `@angular/localize`schematics.
- Remove `commitizen` and configuration.

## README Changes

- Improve wording
- Add new sections:
  - Shove Progress
  - Continuous Integration (CI) Using GitHub Actions
  - Optional Angular Schematics:
    - Angular Material & Angular CDK
    - Server-side & hybrid rendering (SSR)
    - Internationalization (i18n)
    - End to End Testing (e2e)
      - Include Playwright eslint configuration example.
  - Tips & Tricks:
    - Custom Formatting
    - Git Config
    - Oh My Zsh / Oh My Bash / Oh My Posh
    - JetBrains IDEs

joematthews commented:
@replete I based this PR branch off of your PR branch, although I'm not too sure how many of your lines are left except the `# Extreme Angular 19` in the README.md

I squashed all of my changes in to one. I also did a few `git commit --amend` and `git push --force`'s while I was developing the github action for pull requests.

I'll try not to make this a habit, but this PR contains some pretty big changes because I think I'm still trying to solidify what I want extreme-angular to be.

Can you look things over? Your feedback is very much appreciated. Thank you. :pray:

joematthews commented:
Marking this as draft. Going to investigate `prettier-plugin-organize-attributes` and continue to refine the readme.

replete commented:
Okay awesome, I will review this for sure tomorrow evening, in around 18hrs time

joematthews commented:
@replete & @hkbertoson 

Will you please check out the branch, read thru the [new README.md](https://github.com/joematthews/extreme-angular/tree/update-to-angular-19-with-rework), and run the new scripts listed in each section?

You may need to delete the `update-to-angular-19-with-rework` branch and check out a fresh copy if you've pulled it down before. (I may have done a --force push at some point 😅)

Also, can I get a vibe check? Do you think the changes are headed in the right direction?

joematthews commented:
FYI, these messages will _should_ go away after the project is updated to Angular 19.

```sh
....
remote: GitHub found 15 vulnerabilities on joematthews/extreme-angular's default branch (6 high, 5 moderate, 4 low). To find out more, visit:
remote:      https://github.com/joematthews/extreme-angular/security/dependabot
...
```

replete commented:
Hi @joematthews, I've pulled and tried things out. This looks good ✅

Just a couple of nits really:
- Maybe drop some of the optional recommended `extensions.json`, e.g. the github ones, intellicode - not really essential to the stack
- mention the enabled auto save vscode probably

replete commented:
Out of interest, have you done a11y testing with angular? I was wondering how axe for instance works with angular. Pity they didn't adopt Playwright, but that's probably GOOG not wanting too much MSFT in their stack I guess

joematthews commented:
@replete @hkbertoson Alright. Apologies for the constant changes. I did a thrice over on the README.md honing grammar and readability. I think it is in a good place. I look forward to your reviews!

joematthews commented:
@replete thank you!

>     * Maybe drop some of the optional recommended `extensions.json`, e.g. the github ones, intellicode - not really essential to the stack

This is a good recommendation. I personally have always just installed intellicode out of habit since switching from sublimetext many years ago (because I thought it was neat at the time), and now that I recall, I've just ignored the suggestions more often than not. I'll push that change to the `extensions.json` & README.md in a few.

>     * mention the enabled auto save vscode probably

There is a bullet point for this in the [VSCode section](https://github.com/joematthews/extreme-angular/tree/update-to-angular-19-with-rework?tab=readme-ov-file#vscode):

Can this be better worded better?

Maybe change the verbiage of the list text to `The following VSCode settings have been set in .vscode/settings.json:`

> Out of interest, have you done a11y testing with angular? I was wondering how axe for instance works with angular. Pity they didn't adopt Playwright, but that's probably GOOG not wanting too much MSFT in their stack I guess

Unfortunately, I have not had much experience outside of angular-eslint's provided a11y rules. Additional rulesets that do not conflict would be nice. The [eslint-plugin-axe](https://www.npmjs.com/package/eslint-plugin-axe) is horribly out of date and [https://www.npmjs.com/package/eslint-plugin-jsx-a11y](https://www.npmjs.com/package/eslint-plugin-jsx-a11y) appears to be the react community's solution. 

Linting html with ESLint (in my limited experience) has been pretty hit or miss... Prettier does _a lot_ of what eslint html linters would attempt to do. ...but as I'm typing I'm seeing the https://github.com/yeonjuan/html-eslint project which looks promising. I cannot promise anything but I can take a look at this to see if the rules can be tweaked not to conflict with the existing configuration and see if play's nice with IDEs' built in linters.

joematthews commented:
@replete alright, I changed some verbiage in the vscode section. Please let me know if that is more clear.

Did you notice anything, maybe in the Tips & Tricks section you thought was 'too much'?

joematthews commented:
> Pity they didn't adopt Playwright

Thanks for bringing this up. Check out https://angular.dev/tools/cli/end-to-end. Running `ng e2e` now brings up to wizard to choose from multiple options. I think I'll include a reference to this in the Optional Angular Schematics section after I test it out.

joematthews commented:
@replete 

I added a new e2e section in the README.md and added a example eslint config for playwright in the section. I removed references to catppuccin because it's becoming more apparent its irrelevant 😞.

joematthews commented:
Alright, I'm done 'tweaking' for now. @replete Please let me know if you see anything amiss! I'm waiting on some more feedback from a few folks, and then I think I'm going to squash this monstrosity of a branch down to one commit on top of your PR branch and release on Sunday morning.

replete commented:
Hi @joematthews,

This looks really good to me in terms of the tooling and standards laid out. That's the important thing to get right on a new project, to make sure you're doing things right from the start.

I have a couple of suggestions and will push a branch soon to see if you like it

replete commented:
Hi @joematthews 
I made some changes in this new branch in my fork, it's based on this current PR so should replay if you want to bring it into this PR: 
https://github.com/replete/extreme-angular/tree/use-more-esm-in-tooling

The main change is to adopt ESM and drop CommonJS as primary standard in the tooling. I think we're at a place now where we'd all be happier with just one flavour of modules in the JS ecosystem, and its mostly there. 

Changes in branch:
- set `"type": "module"` in package.json - this means that the `.js` extension will be assumed ESM. If we really need commonJS, we can use `.cjs`. `.mjs` is only required in projects where type:module is not set in package.json. I think this is best practice nowadays as commonjs becomes legacy.
- convert and rename `eslint.config.js` to ESM:
    - The rename to .js is parity with type:module being added
    - I've also renamed some of the imported objects for more precision on what they are - this kind of thing really helps when projects grow. I would have pulled `{ configs as esConfigs }` for the eslint/js, but its apparantly a commonJS module, which is really weird, so we can't do that without eslint falling apart. A compromise
- add `.js` single quote rule to `.editorconfig` and reformatted eslint config file for consistency
- adopted `globals.node` etc instead of `{ node }` for the globals packages - `node` `prettier` aren't great names in terms of potential conflicts, and don't represent what they actually are. We could have done `{ node as nodeGlobal, ... } etc, but would be unnecessarily terse, and globals.whatever is already very clear what it is. This is a principle I've found really important in shared codebases - name the thing _very_ closely to what it is

See what you think, but this is something I'd want in any modern project, make everything ESM and use .cjs if I really have to. Most popular toolings now support ESM, and that's the standard now so why not.

The other thing that bugs me that as a user of this package, is the 2-space indent on JS/TS source, I personally find it less readable, but of course its subjective, but it doesn't seem as good as 4 to me generally. Again, not a big deal for an end user would just change it and fix all, and if its good enough for airbnb. Well, it is an interesting point though, I see the benefit of more code being visible, but it just seems for most people harder to read? I wonder if there are any surveys on this... either way genuinely not a big deal for me

joematthews commented:
I couldn't agree more. I only stuck with CommonJS for eslint.ocnfig.js because that's what angular-eslint produced -- but as I got more and more annoyed when incorporating libraries in to the config.

Switching to ECMAScript modules will also make it possible for [prettier-plugin-organize-imports](https://www.npmjs.com/package/prettier-plugin-organize-imports) to do it's magic in config files too.

I'll checkout your branch very soon.

replete commented:
I did notice the automatic import organizing after I ported eslint config to ESM. Very nice!

If these changes make it in will need an update in README.md for the playwright code sample, have to pop out for the rest of the day but will have a look what's happening tomorrow.

joematthews commented:
> The other thing that bugs me that as a user of this package, is the 2-space indent on JS/TS source.

I get it. Hot topic. I don't disagree about the readability. I am used to it at this point.

I think it's best to keep 2 character indentation for these reasons:

- Angular documentation and the angular project itself uses 2 spaces
- Prettier defaults to 2 spaces
- 4 spaces can look pretty bad in JS/TS sometimes with an 80 character line limit
- [2 space indention is traditional for various reasons](https://www.google.com/search?client=firefox-b-1-d&q=2+or+4+space+indention+for+typescript)

Aside from all that, for me personally, I really like the ability to do side-by-side editors and in my experience that really sucks in JS/TS unless indentation 2 and the line width is 80.

All that said, I have have become quite used to Prettier's output for files, but I do remember saying 'ewww' out loud the first time I formatted all files in a project hah.

Changing the indentation and line width is pretty easy. In the .editorconfig set the desired line width in the overrides for the desired file types:

```ini
[*.ts]
quote_type = single
indent_size = 4
ij_typescript_use_double_quotes = false
```

And then in the .prettierrc.json file set the tabWidth & printWidth to your desired values. Again, this will make side-by-side editors, on anything except ultra-wide monitors, a pain to use:

```json
{
    "tabWidth": 4,
    "printWidth": 120,
    "htmlWhitespaceSensitivity": "ignore",
    "plugins": [
        "prettier-plugin-sh",
        "prettier-plugin-css-order",
        "prettier-plugin-organize-imports"
    ]
}
```
And then review, and commit the results.

Should I include these instructions in the tips and tricks? :thinking:

replete commented:
Hi Joe, I probably didn't need to think out loud there, I will adapt and I suspect end up liking it...

replete commented:
What would be really extreme, would be forcing tabs, which would let anyone display their code however they want 🤣

After contracting for some years, I learned that tabs are not a hill worth dying on...

joematthews commented:
> Hi Joe, I probably didn't need to think out loud there, I did notice the angular docs after writing the comment. I will adapt and I suspect end up liking it...

No, it was important to bring up! It has come up in every team I've been on when setting up Prettier. Some teams have opted for longer line widths and using double quotes in TS files. But, I will admit, I've talked them all out of 4 spaces! haha

Not using tabs was a missed opportunity. But there can be conflicts between editors that expect spaces. The .editorconfig file can help a lot, basically instructing editor's default behavior (if they have a plugin for it).

If you want to try out tabs, you can look at the [useTabs](https://prettier.io/docs/en/options.html#tabs) option for Prettier. This will allow the pre-commit to reformat changes correctly with tabs before commit (no matter what other devs are doing in their editor).

That's the beauty of prettier --write in the pre-commit.  Doesn't help if they do --no-verify though on commit :(  That's where formatting checks in pre-push and in CI can help.

joematthews commented:
@replete I merged in your changes. Thank you very much!

The only fix I did was reverting the files proprety in the jasmine section to `'src/**/*.spec.ts'` so they do not conflict with potential `'e2e/**/*.spec.ts'` files if they opt for playwright. (this was something I corrected for earlier in my barrage of pushes lol)

replete commented:
Awesome, it's looking tidy! I'll test the new setup again when it all lands in main with a new project next week. 

I'm churning through trying to figure out best practices for all the 'new' Angular stuff – all the books I'm reading on this are still mostly for Angular 17 and it seems to me like zoneless + signals are a direction I should be headed, but there'll be some lag I guess with best practices so it feels like early days with all that.

joematthews commented:
@replete yeah, I'm pretty new to both zoneless and signals myself.

I have a few projects coming up and I think I will keep zonejs and use signals more (or exclusively if possible).

In the [zoneless docs](https://angular.dev/guide/experimental/zoneless) it mentions using [OnPush](https://angular.dev/best-practices/skipping-subtrees#using-onpush) to be safe. In the new eslint.config.js I'll probably be removing this line in my projects, so it enforces onpush for every component:

```js
      '@angular-eslint/prefer-on-push-component-change-detection': 'off',
```

I don't want to make this default quite yet in extreme-angular tho :P But it's a hella cool rule to have available.

joematthews commented:
@replete I fixed some last minute things I noticed while review the diff and creating the change notes. I'm going to go ahead and pull the trigger on the release.
